### PR TITLE
@truffle/hdwallet provider: Provider auto detector

### PR DESCRIPTION
**Pull request by MrRefactoring**
Determines whether to use `WebsocketProvider` or `HttpProvider` depending on the URL

### CODE REVIEW

**Comment by eggplantzzz** [comment]
This seems pretty cool to me, thanks @MrRefactoring!

**Comment by eggplantzzz** [comment]
@MrRefactoring Any chance you can make the requested edits? Then we can get this PR through.

File: packages/hdwallet-provider/package.json [hunk 1]
```diff
@@ -32,13 +32,13 @@
...
-    "@types/ethereumjs-util": "^5.2.0",
+    "@types/ethereumjs-util": "^6.1.0",
     "@types/mocha": "^5.2.7",
     "@types/web3-provider-engine": "^14.0.0",
     "ganache-core": "2.8.0",
-    "mocha": "6.2.0",
+    "mocha": "^6.2.2",
     "ts-node": "^8.4.1",
-    "typescript": "^3.6.3"
+    "typescript": "^3.7.2"
```
[the remark this answers is not in the document]
Comment (MrRefactoring) [reply packages/hdwallet-provider/package.json]: Done!

**Comment by MrRefactoring** [comment]
@eggplantzzz done! Thx

**Comment by eggplantzzz** [comment]
Sorry for the delay @MrRefactoring, let me get one more approval from someone else on the team. Then if everything is good and you can regenerate the `yarn.lock` we'll get it merged!

File: yarn.lock [hunk 2]
```diff
@@ -14928,9 +14938,9 @@ typescript@3.6.x:
   integrity sha512-unoCll1+l+YK4i4F8f22TaNVPRHcD9PA3yCuZ8g5e0qGqlVlJ/8FSateOLLSagn+Yg5+ZwuPkL8LFUc0Jcvksg==
 
 typescript@^3.6.2, typescript@^3.6.3:
-  version "3.7.3"
-  resolved "https://registry.yarnpkg.com/typescript/-/typescript-3.7.3.tgz#b36840668a16458a7025b9eabfad11b66ab85c69"
-  integrity sha512-Mcr/Qk7hXqFBXMN7p7Lusj1ktCBydylfQM/FZCk5glCNQJrCUKPkMHdo9R0MTFWsC/4kPFvDS0fDPvukfCkFsw==
+  version "3.7.4"
+  resolved "https://registry.yarnpkg.com/typescript/-/typescript-3.7.4.tgz#1743a5ec5fef6a1fa9f3e4708e33c81c73876c19"
+  integrity sha512-A25xv5XCtarLwXpcDNZzCGvW2D1S3/bACratYBx2sax8PefsFhlYmkQicKHvpYflFS8if4zne5zT5kpJ7pzuvw==
```
Comment (gnidan) [inline yarn.lock]: Looks like these got left in!

**Pull request merged**